### PR TITLE
[sources] Thread a should_halt flag through the source pipeline

### PR DESCRIPTION
**Pull request by bkirwi**
This shifts the halt-on-snapshot error from the postgres source to the healthcheck operator, allowing us to ensure a stall is reported before shutdown.

This is objectively pretty sketchy, but it is:
- More reliable than sleeping in the source and hoping for the best.
- Much easier than inventing a graceful shutdown protocol for the source, which is an ongoing discussion point.

### Motivation

Known issue: https://github.com/MaterializeInc/materialize/issues/16581

### Tips for reviewer

I've also updated the tests to not avoid running during the snapshot, but of course it's a bit hard to ensure that they _do_ run during the snapshot. The tests hould at least verify that we're not breaking anything else in the source pipeline with this refactor.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by aljoscha — APPROVED** [review]
This doesn't look _too_ hairy! But I'd also want @petrosagg  to have a look before we merge.

File: src/storage/src/source/postgres.rs
Comment (petrosagg) [inline src/storage/src/source/postgres.rs:452]: That looks strange, can't we return from the function instead?

**Review by petrosagg — APPROVED** [review]
I think for me sleeping would be preferable, because at least it doesn't lift an unwanted behaviour (halting) to framework code and it's just some bad thing that only the pg source does.

If you really want to merge this it's also fine as long as we don't start using this field for anything else. Maybe we could name it `should_halt_do_not_use` or something, up to you.

File: src/storage/src/source/postgres.rs
Comment (bkirwi) [reply src/storage/src/source/postgres.rs:452]: IDK, maybe, but I'm skeptical!

IIUC returning here will cause the downstream poller to return `Finished`, and I'm worried about that eg. advancing the output collection's upper to the empty frontier. On the other hand, just having the source be `Pending` until the halt is triggered seems much safer.

**Comment by bkirwi** [comment]
I don't really have an opinion about the sleep vs. this `halt_requested` thing. (Well, I now like this one better since I have an approved PR for it. 😅)

**Comment by petrosagg** [comment]
Let's just merge it and we'll deal with halting later

**Pull request merged**
